build.sc

```diff
@@ -0,0 +1,239 @@
+    os.proc("sphinx-build", "-b", "html", millSourcePath / os.up / "docs" / "manual", T.dest).call()
+  }
+
+  def launch(args: String*) = T.command {
+    os.proc("java", args, "-jar", localDeploy().path / "ArtOfIllusion.jar").call()
+  }
+
+  def stage = T {
+    os.makeDir.all(T.dest / "lib")
+    os.makeDir.all(T.dest / "Plugins")
+    os.makeDir.all(T.dest / "Textures and Materials")
+    os.makeDir.all(T.dest / "Scripts" / "Tools")
+    os.makeDir.all(T.dest / "Scripts" / "Startup")
+    os.makeDir.all(T.dest / "Scripts" / "Objects")
+
+    ArtOfIllusion.upstreamAssemblyClasspath()
+      .map(_.path)
+      .filter(p => os.exists(p) && os.isFile(p))
+      .iterator
+      .foreach(p => os.copy(p, T.dest / "lib" / p.last))
+
+    os.copy(ArtOfIllusion.jar().path, T.dest / "ArtOfIllusion.jar")
+
+    os.copy(Filters.jar().path, T.dest / "Plugins" / Filters.artifactName().concat(".jar"))
+    os.copy(OSSpecific.jar().path, T.dest / "Plugins" / OSSpecific.artifactName().concat(".jar"))
+    os.copy(Renderers.jar().path, T.dest / "Plugins" / Renderers.artifactName().concat(".jar"))
+    os.copy(Tools.jar().path, T.dest / "Plugins" / Tools.artifactName().concat(".jar"))
+    os.copy(Translators.jar().path, T.dest / "Plugins" / Translators.artifactName().concat(".jar"))
+    mill.modules.Util.download("http://aoisp.sourceforge.net/AoIRepository/Plugins/SPManager/SPManager-3_0.jar",
+      os.rel / "Plugins" / "SPManager.jar")
+    mill.modules.Util.download("http://aoisp.sourceforge.net/AoIRepository/Plugins/SPManager/PostInstall-3_0.jar",
+      os.rel / "Plugins" / "PostInstall.jar")
+    PathRef(T.dest)
+  }
+
+  def localDeploy = T.persistent {
+    os.copy(stage().path, T.dest, replaceExisting = true, mergeFolders = true)
+    val osName = System.getProperty("os.name").toLowerCase()
+    val natives = osName match {
+      case x if osName.contains("linux") => Cache.joglLinuxNatives()
+      case x if osName.contains("windows") => Cache.joglWindowsNatives()
+      case x if osName.contains("mac") | osName.contains("osx") => Cache.joglMacOSXNatives()
+    }
+
+    natives
+      .map(_.path)
+      .iterator
+      .foreach(p => os.copy(p, T.dest / "lib" / p.last, replaceExisting = true))
+
+    PathRef(T.dest)
+  }
+}
+
+/* Because if we leave this in Common, all the downloaded files will be downloaded/cached
+ * per Module object.
+ */
+object Cache extends Module{
+  def unmanagedFiles = T {
+    Agg(
+      downloadFile(
+        "https://github.com/blackears/svgSalamander/releases/download/v1.1.3/svgSalamander-1.1.3.jar"),
+      downloadFile(
+        "https://github.com/beanshell/beanshell/releases/download/2.1.0/bsh-2.1.0.jar"),
+      download("https://maven.jzy3d.org/releases/org/jogamp/gluegen/gluegen-rt/v2.4.0-rc4/gluegen-rt-v2.4.0-rc4.jar",
+        os.rel / "gluegen-rt.jar"),
+      download("https://maven.jzy3d.org/releases/org/jogamp/jogl/jogl-all/v2.4.0-rc4/jogl-all-v2.4.0-rc4.jar",
+        os.rel / "jogl-all.jar")
```

Comment:
You're manually downloading lots of dependencies from a Maven repository. Why don't you configure the repository via `repositoryTask` and use `ivyDeps` or `runIvyDeps` instead?

Reply:
Any thoughts on how to shade jar names *outside* of the `assembly` target? I've found `Assembly.getShadedClasspath`, but have no idea what to do with a `Generator` - I don't even understand the type signature.

Reply:
I have no idea what you mean with "shade jar names".

Reply:
Now I understand. The repo isn't really broken, but you need jars without the version number in the filename. How about replacing the jars names after they are resolved? e.g. in `compileClasspath` or `runClasspath`, you could just map over the super target result and copy those that need modification into `T.dest` with the new filename and use these instead.

Reply:
Thanks for the hint. Since this only really matters when *deploying*, I modified the names while copying the files into the deployment assembly. This should only last until we upgrade to a newer version of the library, or replace it altogether.